src/passes/StringLowering.cpp

```diff
@@ -175,6 +176,61 @@ struct StringGathering : public Pass {
   }
 };
 
+struct StringLowering : public StringGathering {
+  void run(Module* module) override {
+    if (!module->features.has(FeatureSet::Strings)) {
+      return;
+    }
+
+    // First, run the gathering operation so all string.consts are in one place.
+    StringGathering::run(module);
+
+    // Lower the string.const globals into imports.
+    makeImports(module);
+
+    // Remove all HeapType::string etc. in favor of externref.
+    updateTypes(module);
+
+    // Disable the feature here after we lowered everything away.
+    module->features.disable(FeatureSet::Strings);
+  }
+
+  void makeImports(Module* module) {
+    Index importIndex = 0;
+    json::Value stringArray;
+    stringArray.setArray();
+    std::vector<Name> importedStrings;
+    for (auto& global : module->globals) {
+      if (global->init) {
+        if (auto* c = global->init->dynCast<StringConst>()) {
+          global->module = "string.const";
+          global->base = std::to_string(importIndex);
+          importIndex++;
+          global->init = nullptr;
+
+          auto str = json::Value::make(std::string(c->string.str).c_str());
+          stringArray.push_back(str);
+        }
+      }
+    }
+
+    // Add a custom section with the JSON.
+    std::stringstream stream;
+    stringArray.stringify(stream);
+    auto str = stream.str();
+    auto vec = std::vector<char>(str.begin(), str.end());
+    module->customSections.emplace_back(
+      CustomSection{"string.consts", std::move(vec)});
+  }
+
+  void updateTypes(Module* module) {
+    TypeMapper::TypeUpdates updates;
+    updates[HeapType::string] = HeapType::ext;
```

Comment:
If a module ever uses the fact that `string <: any`, this is going to cause problems, but that's probably ok for experimentation.

Reply:
Good point. If that is ever an issue it seems like we'd need to internalize/externalize in a lot of places...

Reply:
This is why I think we should make our internal string type be a subtype of `extern` instead of `any`, but of course that would break compatibility with the stringref proposal. This is a change we should make once we don't need to support stringref directly anymore and can get away with only supporting imported strings. Hopefully that will be soon?

Reply:
I see, yeah, that could help. But implementing a not-quite-stringref has downsides too. I'm not sure what's best in the long term, but we don't need to decide now.